File: 0516-longest-palindromic-subsequence/0516-longest-palindromic-subsequence.java
class Solution {
    private int lcs(String s,String t){
        int n = s.length();
        int m = t.length();
        
        int dp[][] = new int[n+1][m+1];
        for(int row[] : dp) Arrays.fill(row,-1);
        
        for(int i = 0;i <= n; i++){
            dp[i][0] = 0;
        }
        for(int j = 0;j <= m; j++){
            dp[0][j] = 0;
        }
        for(int i = 1; i <= n; i++){
            for(int j = 1; j <= m; j++){
                if(s.charAt(i-1) == t.charAt(j-1)){
                    dp[i][j] = 1 + dp[i-1][j-1];
                }else{
                    dp[i][j] = Math.max(dp[i-1][j], dp[i][j-1]);
                }
            }
        }
        return dp[n][m];
    }
    public int longestPalindromeSubseq(String s) {
        String t = s;
        String ss=new StringBuilder(s).reverse().toString();
        return lcs(ss, t);
    }
}

File: 0516-longest-palindromic-subsequence/0516-longest-palindromic-subsequence.cpp
class Solution {
private:
    int longestCommonSubsequence(string s1, string s2) {
        int n = s1.size();
        int m = s2.size();
        // vector<vector<int>> dp(n+1,vector<int>(m+1,0));
        vector<int> prev(m+1,0),cur(m+1,0);
        
        for(int j = 0; j < m; j++) {
            prev[j] = 0;
        }
        
        for(int ind1 = 1; ind1 <= n; ind1++) {
            for(int ind2 = 1; ind2 <= m; ind2++) {
                if(s1[ind1-1] == s2[ind2-1]) {
                    cur[ind2] = 1 + prev[ind2-1];
                }
                else  cur[ind2] = max(prev[ind2] , cur[ind2-1]);
            }
            prev = cur;
        }
        return prev[m];
    }
public:
    int longestPalindromeSubseq(string s1) {
        int n = s1.size();
        string s2 = s1;
        reverse(s2.begin(),s2.end());
        return longestCommonSubsequence(s1,s2);
    }
};